### ISSUE TITLE
Is there an ability to chain multiple validation methods returning ErrorOr<T>?

### ISSUE THREAD

**Issue by mickvikt** [amantinband/error-or#75]
Having a validation depending on multiple checks, is constructing a list of errors the only way to return a failure result for multiple fails?
Having a Bind extension method or something titled more idiomatic to C# would be nice, since this would enable to chain multiple operations in a single flow.

**Comment by apfohl** [comment]
You could even leverage the powers of Linq and implement `ErrorOr<TResult> SelectMany<T, TCollection, TResult>(this ErrorOr<T> errorOr, Func<T, ErrorOr<TCollection>> collection, Func<T, TCollection, TResult> selector)`.

That would allow for things like that:

```csharp
from user in userRepository.Lookup("Andreas")
from deletedUser in userRepository(user)
select deletedUser
```

That would enable some sort of railway oriented programming style. You could even do it for the `async` variant.

What do you think @amantinband? It would be fun to build this addition.

**Comment by apfohl** [comment]
Here you go: https://github.com/emzdyhravemzdy/LinqErrorOr

Someone already did it.

**Comment by amantinband** [comment]
Can you guys take a look at #77 and let me know what you think? It's inspired by #31 with some tweaks (cc @lucafabbri)

tl;dr:

## Chain

```cs
 [Fact]
    public void ChainErrorOrs_WhenHasValue_ShouldInvokeNextInChain()
    {
        // Arrange
        ErrorOr<Person> errorOrPerson = new Person("Amichai");

        static ErrorOr<string> GetName(Person person) => person.Name;
        static ErrorOr<Person> CreatePersonFromName(string name) => new Person(name);

        // Act
        ErrorOr<Person> result = errorOrPerson
            .Chain(person => GetName(person))
            .Chain(name => CreatePersonFromName(name));

        // Assert
        result.IsError.Should().BeFalse();
        result.Value.Should().BeEquivalentTo(errorOrPerson.Value);
    }

    [Fact]
    public void ChainErrorOrs_WhenHasError_ShouldReturnErrors()
    {
        // Arrange
        ErrorOr<Person> errorOrPerson = Error.NotFound();

        static ErrorOr<string> GetName(Person person) => person.Name;
        static ErrorOr<Person> CreatePersonFromName(string name) => new Person(name);

        // Act
        ErrorOr<Person> result = errorOrPerson
            .Chain(person => GetName(person))
            .Chain(name => CreatePersonFromName(name));

        // Assert
        result.IsError.Should().BeTrue();
        result.FirstError.Should().BeEquivalentTo(Error.NotFound());
    }
```

## ChainAsync

```cs
    [Fact]
    public async Task ChainErrorOrsAsync_WhenStateIsValue_ShouldInvokeNextInChain()
    {
        // Arrange
        ErrorOr<Person> errorOrPerson = new Person("Amichai");

        static Task<ErrorOr<string>> GetNameAsync(Person person) => Task.FromResult(ErrorOrFactory.From(person.Name));
        static Task<ErrorOr<Person>> CreatePersonFromNameAsync(string name) => Task.FromResult(ErrorOrFactory.From(new Person(name)));

        // Act
        var result = await errorOrPerson
            .ChainAsync(person => GetNameAsync(person))
            .ChainAsync(name => CreatePersonFromNameAsync(name))
            .ChainAsync(person => GetNameAsync(person))
            .ChainAsync(name => CreatePersonFromNameAsync(name));

        // Assert
        result.IsError.Should().BeFalse();
        result.Value.Should().BeEquivalentTo(errorOrPerson.Value);
    }

    [Fact]
    public async Task ChainErrorOrsAsync_WhenStateIsError_ShouldReturnErrors()
    {
        // Arrange
        ErrorOr<Person> errorOrPerson = Error.NotFound();

        static Task<ErrorOr<string>> GetNameAsync(Person person) => Task.FromResult(ErrorOrFactory.From(person.Name));
        static Task<ErrorOr<Person>> CreatePersonFromNameAsync(string name) => Task.FromResult(ErrorOrFactory.From(new Person(name)));

        // Act
        var result = await errorOrPerson
            .ChainAsync(person => GetNameAsync(person))
            .ChainAsync(name => CreatePersonFromNameAsync(name))
            .ChainAsync(person => GetNameAsync(person))
            .ChainAsync(name => CreatePersonFromNameAsync(name));

        // Assert
        result.IsError.Should().BeTrue();
        result.FirstError.Should().BeEquivalentTo(Error.NotFound());
    }
```

**Comment by apfohl** [comment]
@amantinband Yes that looks great.

I must apologize, I couldn't wait and build this: https://github.com/apfohl/ErrorOrExtensions/.

Once you support it natively I can remove it again 😅.

I would love to have the LINQ expression syntax though... 😉

**Comment by amantinband** [comment]
Cool! I just don't love the naming.. `Chain` or `Select` both feel a bit off. Looking at your repo, I really like the`OrElse` idea and the naming is very intuitive.

This I don't like:
```cs
    [HttpGet]
    public Task<IActionResult> Edit(Guid id) =>
        booksRepository.Lookup(id)
            .Chain(book => book)
            .Chain(book => View(new BooksEditViewModel(book)))
            .OrElse(StatusCode(500, "An unexpected error occurred on the server."));
```

Perhaps something like following:
```cs
    [HttpGet]
    public Task<IActionResult> Edit(Guid id) =>
        booksRepository.Lookup(id)
            .ThenIfSuccess(book => book)
            .ThenIfSuccess(book => View(new BooksEditViewModel(book)))
            .OrElse(StatusCode(500, "An unexpected error occurred on the server."));
```

or

```cs
    [HttpGet]
    public Task<IActionResult> Edit(Guid id) =>
        booksRepository.Lookup(id)
            .ContinueWith(book => book)
            .ContinueWith(book => View(new BooksEditViewModel(book)))
            .Breakwith(StatusCode(500, "An unexpected error occurred on the server."));
```

Something around these lines I think can be a really good addition

**Comment by mickvikt** [comment]
Regarding naming, I think `.Then()/.OrElse()` seem to be both understandable and succinct, i.e.: 

```
    [HttpGet]
    public Task<IActionResult> Edit(Guid id) =>
        booksRepository.Lookup(id)
            .Then(book => Transform(book))
            .Then(book => View(new BooksEditViewModel(book)))
            .OrElse(StatusCode(500, "An unexpected error occurred on the server."));
```

A couple of chained `.Then`s seem to implicitly suggest that all of the checks/actions have completed successfully, so I don't really se the point in naming `.ThenIfSuccess`

**Comment by apfohl** [comment]
I agree with @mickvikt. Naming it `Then()` makes it pretty clear what happened. It relates a little bit to a `Promise` in JavaScipt. Even though it is for async processing.

`OrElse` should have three implementations though:

```csharp
public static T OrElse<T>(this ErrorOr<T> errorOr, T orElse) =>
    errorOr.Match(value => value, _ => orElse);

public static T OrElse<T>(this ErrorOr<T> errorOr, Func<T> orElse) =>
    errorOr.Match(value => value, _ => orElse());

public static T OrElse<T>(this ErrorOr<T> errorOr, Func<List<Error>, T> orElse) =>
    errorOr.Match(value => value, orElse);
```

**Comment by amantinband** [comment]
yeah that's better but still doesn't feel perfect IMO.

Also, maybe we should just drop it to `Else`:

```cs
    [HttpGet]
    public Task<IActionResult> Edit(Guid id) =>
        booksRepository.Lookup(id)
            .Then(book => Transform(book))
            .Then(book => View(new BooksEditViewModel(book)))
            .Else(StatusCode(500, "An unexpected error occurred on the server."));
```

Regarding the 3 methods - I guess that makes sense for fluentness. I'll have the `Then` method behave the same so we can do the following:

```cs
    [HttpGet]
    public Task<IActionResult> ReturnOkIfExists(Guid id) =>
        booksRepository.Exists(id)
            .Then(Ok())
            .Else(NotFound());
```

Sounds good? Let me know what you think

**Comment by apfohl** [comment]
Yes, ´Then´ and ´Else´ is a good idea. People are struggling forever with naming these methods.

Also, I wouldn't think to hard about the naming of ´Else´ because it is basically just a your ´Match´ when you combine it with ´Then´. But ´Match´ takes a function returning ´T´ and not `ErrorOr<T>`. So ´Match´ is actually a combination of ´Map´ (´Select´ in my case) and ´Else´. I took the name ´OrElse´ from the Java ´Optional<T>´ implementation.

**Comment by amantinband** [comment]
This is where we're at:

### `Then` / `ThenAsync`

Multiple methods that return `ErrorOr<T>` can be chained as follows:

```csharp
static ErrorOr<string> ConvertToString(int num) => num.ToString();
static ErrorOr<int> ConvertToInt(string str) => int.Parse(str);

ErrorOr<string> errorOrString = "5";

ErrorOr<string> result = errorOrString
    .Then(str => ConvertToInt(str))
    .Then(num => ConvertToString(num))
    .Then(str => ConvertToInt(str))
    .Then(num => ConvertToString(num));
```

```csharp
static ErrorOr<string> ConvertToString(int num) => num.ToString();
static Task<ErrorOr<string>> ConvertToStringAsync(int num) => Task.FromResult(ErrorOrFactory.From(num.ToString()));
static Task<ErrorOr<int>> ConvertToIntAsync(string str) => Task.FromResult(ErrorOrFactory.From(int.Parse(str)));

ErrorOr<string> errorOrString = "5";

ErrorOr<string> result = await errorOrString
    .ThenAsync(str => ConvertToIntAsync(str))
    .ThenAsync(num => ConvertToStringAsync(num))
    .ThenAsync(str => ConvertToIntAsync(str))
    .ThenAsync(num => ConvertToStringAsync(num));

// mixing `ThenAsync` and `Then`
ErrorOr<string> result = await errorOrString
    .ThenAsync(str => ConvertToIntAsync(str))
    .Then(num => ConvertToString(num))
    .ThenAsync(str => ConvertToIntAsync(str))
    .Then(num => ConvertToString(num));
```

If any of the methods return an error, the chain will break and the errors will be returned.

### `Else` / `ElseAsync`

The `Else` / `ElseAsync` methods can be used to specify a fallback value in case the state is error anywhere in the chain.

```csharp
// ignoring the errors
string result = errorOrString
    .Then(str => ConvertToInt(str))
    .Then(num => ConvertToString(num))
    .Else("fallback value");

// using the errors
string result = errorOrString
    .Then(str => ConvertToInt(str))
    .Then(num => ConvertToString(num))
    .Else(errors => $"{errors.Count} errors occurred.");
```

```csharp
// ignoring the errors
string result = await errorOrString
    .ThenAsync(str => ConvertToInt(str))
    .ThenAsync(num => ConvertToString(num))
    .ElseAsync(Task.FromResult("fallback value"));

// using the errors
string result = await errorOrString
    .ThenAsync(str => ConvertToInt(str))
    .ThenAsync(num => ConvertToString(num))
    .ElseAsync(errors => Task.FromResult($"{errors.Count} errors occurred."));
```

@apfohl I ended up only adding two methods, since the `Func<T>` can be achieved by calling the `Func<List<Error>, T>` version like this:
```
ErrorOr<int> result = errorOrString
    .Then(_ => 5)
```

**Comment by amantinband** [comment]
I'd appreciate if some of you can take a look at the PR and make sure I didn't miss something obvious.

I'm good to go with this version unless someone has good suggestions 🤙

**Comment by apfohl** [comment]
I'll give it a review.

**Comment by apfohl** [comment]
> @apfohl I ended up only adding two methods, since the `Func<T>` can be achieved by calling the `Func<List<Error>, T>` version like this:
> 
> ```
> ErrorOr<int> result = errorOrString
>     .Then(_ => 5)
> ```

That's fair.
